Add waitcounters for torch.compile subprocess pool

c00w opened the pull request:
Summary:
This ads waitcounter for whether or not the pool is running, as well as if we
are running jobs.

This also ads waitcounters for the first job within a pool. First job and running are working correctly. The job waitcounter seems to either be detecting a leak of a job, or is broken subtly.

Test Plan:
We've tested this internally and see valid ods metrics.

Note that we may be leaking jobs, or the job logic may not be handling an exception correctly.

Differential Revision: D83705931




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -192,9 +194,13 @@ def __init__(
         self.job_id_count = itertools.count()
 
         self.running = True
+        self.running_waitcounter =  _WaitCounter( "pytorch.wait_counter.subproc_pool.running").guard()
```
What is "running" supposed to capture? Because I don't think the ProcessPoolExecutor inside the subproc is running until a wakeup call or the first just is submitted.

But as written, it seems to be capturing the interval when the pool object is initialized to when the process exits. But that just means we compiled at least one Triton kernel.

c00w replied:
Happy to rename this - Maybe created ? It's intended to be all time while the parent object exists and it's own internal running variable isn't set to false (i.e. we're not exiting).

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -192,9 +194,13 @@ def __init__(
         self.running = True
+        self.running_waitcounter =  _WaitCounter( "pytorch.wait_counter.subproc_pool.running").guard()
+        self.running_waitcounter.__enter__()
+        self.firstjob = True
```
Can you also initialize the other "firstjob" related members here? Ideally a typechecker would complain that you could be accessing firstjob_id before it's initialized, but I guess Python.

Maybe you can also combine firstjob and firstjob_id into one variable. Like if firstjob_id is None, that means first job.

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -192,9 +194,13 @@ def __init__(
 
         self.futures_lock = threading.Lock()
         self.pending_futures: dict[int, Future[Any]] = {}
+        self.pending_waitcounters: dict[int, Any] = {}
```
What is pending_waitcounters supposed to capture? The interval when the pool is doing work? If so I wonder if we could simplify and just have one waitcounter that entered/exited when there are no jobs (i.e., when len(pending_futures) == 0). Does that do the right thing?

c00w replied:
Yes - it's trying to capture the time when we have any work running. Note that waitcounters are "reentrant" so you'll get meaningful data out of them if we have a waitcouner per job (i.e. duration_avg is how long each job took, and call_count is the number of jobs we are processing per minute).

No strong objection to a inverted counter, but it's actually losing information we get with the pending waitcounters.

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -263,20 +277,31 @@ def _read_thread(self) -> None:
 
     def quiesce(self) -> None:
         self._send(MsgHeader.QUIESCE)
+        self.quiesce_waitcounter =  _WaitCounter( "pytorch.wait_counter.subproc_pool.running").guard()
```
So you want to capture the interval the pool is shutdown. This probably needs to be initialized earlier too. Wakeup would be called before quiesce.

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -263,20 +277,31 @@ def _read_thread(self) -> None:
     def wakeup(self) -> None:
         self._send(MsgHeader.WAKEUP)
+        self.quiesce_waitcounter.__exit__()
+        self.firstjob = True
```
Unfortunately, this placement for firstjob=True isn't robust. I don't know if there are any codepaths that neglect to call wakeup, but it's is optional. If you submit to the pool before a wakeup, the pool will wake itself up. Probably set it to True in quiesce? If there's a race, I'm not sure that even works correctly though :/

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -237,6 +250,7 @@ def _read_thread(self) -> None:
                 if self.running:
                     log.warning("SubprocPool unclean exit")
                     self.running = False
+                    self.running_waitcounter.__exit__()
```
Probably this can rely on the one in shutdown? See comment there.

c00w replied:
So we're tieing this waitcounter strictly to the running variable - In particular, running is only set to false in two spots, this code XOR the code below. Thus we're exiting in both spots.

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -263,20 +277,31 @@ def _read_thread(self) -> None:
                 if not self.running:
                     return
                 self.running = False
+                self.running_waitcounter.__exit__()
```
I guess I don't know what this WC is capturing. Is it safe to exit waitcounters multiple times? If so, maybe you can move this above the check for self.running and call it unconditionally.

c00w commented:
Still working on fixing the typing issue, but should be reviewable.

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -193,9 +194,26 @@ def __init__(
+        self.running_waitcounter.__enter__()
+
+        # The quiesce waitcounter indicates when the job is in a quiesced state.
+        self.quiesce_waitcounter: Optional[_WaitCounterTracker] = None
```
I guess this could be slightly confusing; maybe it's fine. The pool actually starts in a quiesced state in the sense that the subprocess's internal pool is not initialized yet. The internal pool gets created on a wakeup() -or- when the first job is submitted.

c00w committed:
Add waitcounters for torch.compile subprocess pool (#164527)

Summary:

This ads waitcounter for whether or not the pool is running, as well as if we
are running jobs.

This also ads waitcounters for the first job within a pool. First job and running are working correctly. The job waitcounter seems to either be detecting a leak of a job, or is broken subtly.

Test Plan:
We've tested this internally and see valid ods metrics.

Note that we may be leaking jobs, or the job logic may not be handling an exception correctly.

Reviewed By: masnesral

Differential Revision: D83705931

c00w commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -212,6 +230,19 @@ def submit(
+                "pytorch.wait_counter.subproc_pool.job"
+            ).guard()
+            self.pending_waitcounters[job_id].__enter__()
+            if self.quiesce_waitcounter:
```
Yeah ... Let me stack a PR to also log from wakeup this (since technically, you're not required to submit before wakeup, and vice versa).
